### PR TITLE
Fixed JAX solver compatibility test

### PR DESCRIPTION
**Pull request by santacodes**
# Description

I caught this while running `nox -s unit` and also in CI [here](https://github.com/pybamm-team/PyBaMM/actions/runs/10285560857/job/28464203323).
Instead of checking the compatibility of the JAX solver the test was checking if JAX solver was present and logging `JAX is not installed`, when it should log something like `JAX is compatible` when the check passes inside the `skipif` marker of pytest.

```bash
SKIPPED [1] tests/unit/test_solvers/test_idaklu_jax.py:91: Both IDAKLU and JAX are available
SKIPPED [1] tests/unit/test_util.py:91: The JAX solver is not installed
```

Fixes # (issue)

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: tests/unit/test_util.py
Comment (kratman) [inline tests/unit/test_util.py:94]: Shouldn't the skip behavior be `!pybamm.have_jax()`? it is testing if jax is compatible.

The current logic:
```
jax installed -> Skip test with a message that jax is not installed
jax not installed -> check if jax is compatible, if so assert true
```

New logic:
```
jax compatible -> Skip test with a message that jax is installed
jax not installed -> run a test that asserts true
```

I don't think the test is working in either setup

**Review by kratman — CHANGES_REQUESTED** [review]

File: tests/unit/test_util.py
Comment (kratman) [reply tests/unit/test_util.py:94]: Maybe a better approach is to remove this test or rework what is actually being tested

Comment (santacodes) [reply tests/unit/test_util.py:94]: I think there are two different tests, one for testing if JAX is present and the other for checking if it is compatible. The presence of JAX and IDAKLU solvers are already being tested and working well, while this function had to test the compatibility and it should probably `assert False` inside the function if the test doesn't skip. So it should probably be like - 
```python 
    @pytest.mark.skipif(
        pybamm.is_jax_compatible(), reason="The JAX solver is compatible"
    )
    def test_is_jax_compatible(self):
        assert False
```
But pre-commit does not want me to assert False in that case, would there be a way to implement this logic? Also, I previously brought up this [here](https://github.com/pybamm-team/PyBaMM/pull/3901#issuecomment-2273383056).
Or probably we could implement like you mentioned 
```python 
    @pytest.mark.skipif(
        !pybamm.is_jax_compatible(), reason="The JAX solver is not compatible"
    )
    def test_is_jax_compatible(self):
        assert True
```

Comment (kratman) [reply tests/unit/test_util.py:94]: I think it is just a bad test. The skipping mechanism is being used for internal logic. The tests should test behavior, not that a module was installed by our installer

Comment (kratman) [reply tests/unit/test_util.py:94]: As a further point a assert false would make tests fail any time jax was not installed

Comment (santacodes) [reply tests/unit/test_util.py:94]: Should I just remove it then?

Comment (santacodes) [reply tests/unit/test_util.py:94]: > I think it is just a bad test. The skipping mechanism is being used for internal logic. The tests should test behavior, not that a module was installed by our installer

Also, there are tests for testing if IDAKLU and JAX are installed in that case maybe we could just make one collective test/validation of some kind for that?

Comment (kratman) [reply tests/unit/test_util.py:94]: There is no sense in keeping tests which don't test what you expect. This test and other similar tests can probably be removed. If it is not testing the behavior of a function, then it is probably a bad test.

Any time you see just a bare `assert True` without an accompanying logic block, then it should raise a red flag.

**Review by kratman — APPROVED** [review]
This works for me

**Comment by agriyakhetarpal** [comment]
I think this came about as a migration done incorrectly in 68f71f278501b8bc670aed1514d4b32a817f42d4? I think we should indeed test that the JAX version is compatible (i.e., `0.4`, since we don't support `0.5`) – which was the earlier behaviour of the test.

**Comment by kratman** [comment]
> I think this came about as a migration done incorrectly in [68f71f2](https://github.com/pybamm-team/PyBaMM/commit/68f71f278501b8bc670aed1514d4b32a817f42d4)? I think we should indeed test that the JAX version is compatible (i.e., `0.4`, since we don't support `0.5`) – which was the earlier behaviour of the test.

While I agree that the change was done poorly in 68f71f278501b8bc670aed1514d4b32a817f42d4, I am still in favor of removing the test. Installing the correct version of jax is the job of the build system, not the unit tests. If we need to check for it, then it should be done in the build process not in the tests after the build is done

**Comment by agriyakhetarpal** [comment]
Okay, fair enough – we have `jax` and `jaxlib` pinned already, and there's a chance the tests will fail elsewhere if someone undoes the pin

**Review by agriyakhetarpal — APPROVED** [review]

**Review by arjxn-py — APPROVED** [review]

**Comment by arjxn-py** [comment]
Oh, just wait for me before you merge.

**Comment by arjxn-py** [comment]
```
@pytest.mark.skipif(not pybamm.have_jax(), reason="JAX is not installed")
def test_is_jax_compatible():
    compatible = pybamm.is_jax_compatible()
    assert compatible
```
This should be the correct test, @santacodes.

File: tests/unit/test_util.py
Comment (arjxn-py) [inline tests/unit/test_util.py:90]:
```suggestion

@pytest.mark.skipif(not pybamm.have_jax(), reason="JAX is not installed")
def test_is_jax_compatible():
    compatible = pybamm.is_jax_compatible()
    assert compatible
```

**Review by arjxn-py — CHANGES_REQUESTED** [review]

File: tests/unit/test_util.py
Comment (kratman) [inline tests/unit/test_util.py:95]: I still think this is unneeded, but if it is being put in, then it needs to be cleaned up
```suggestion
    @pytest.mark.skipif(not pybamm.have_jax(), reason="JAX is not installed")
    def test_is_jax_compatible():
        assert pybamm.is_jax_compatible()
```

**Review by arjxn-py — APPROVED** [review]

File: tests/unit/test_util.py
Comment (kratman) [inline tests/unit/test_util.py:92]:
```suggestion
    def test_is_jax_compatible(self):
```

**Pull request merged**
